docs(kuma-cp): Add summary of health check methods (#688)
kumahq/kuma-website: docs/docs/dev/policies/service-health-probes.md

> # Service Health Probes

Reviewer: This in more `Dataplane Health Probes` or `Inbound Health Probes` than service? Especially as service is very overloaded :)

the environment ([Kubernetes](#kubernetes-probes) or [Universal](#universal-probes)), but it's treated the same way 
regardless of the environment:

- if `health.ready` is true or `health` section is missing - Kuma considers the inbound as healthy and includes it 
  into load balancing set.
- if `health.ready` is false -  Kuma doesn't include the inbound into load balancing set.

Also, `health.ready` is used to compute the status of the Dataplanes and Service. You can see these statuses both in Kuma GUI and Kuma CLI:

- if proxy status is `Offline`, then Dataplane is `Offline`:
- if proxy status is `Online`:
  - if all inbounds are ready then Dataplane is `Online`
  - if all inbounds are not ready then Dataplane is `Offline`
  - if at least one of the inbounds is not ready then Dataplane is `Partially degraded` 
  - if inbound is not ready then it's not included in the load-balancer set which means it doesn't receive the traffic
  - if all inbounds which implement the same service are ready then service is `Online`
  - if all inbounds which implement the same service are not ready then service is `Offline`

>   - if at least one of the inbounds which implement the same service is not ready then service is `Partially degraded`

Reviewer: We probably want to describe the different levels of health.

Each dataplane has multiple inbounds each belonging to a specific service. Each inbound is probed independently.
All inbounds in a mesh that have the same `kuma.io/service` are grouped in what is called a `service`

A service has a status which can take 3 values:

- `Online` if all dataplane inbounds with this service have `health: true`
- `Offline` if all dp inbounds with this service have `health: false`
- `PartiallyDegraded` if any inbounds with this service have `health: false`

A specific dataplane will not be present in the list of endpoints sent to dataplanes trying to communicate with this service if the inbound's matching this service health is false.

There's probably better ways to right this but I feel doing:

1. What is inbound health
2. What is a service
3. What does this mean for advertising to other dataplanes using this service

Helps comprehension.

## Kubernetes probes

> Even if Kubernetes probes are disabled, Kuma takes `pod.status.containerStatuses.ready` in order to fill `dataplane.inbound.health` section.

Reviewer: This is useful info but might be better if it's moved further down (let's first describe what we do :) )

> 

Reviewer: Worth explaining that in K8s Kuma will update Dataplane health according to the status of the containers implementing the service and the status of the sidecar container (Worth pointing to the [k8s doc related to health](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/)).

If sidecar health is false: --> all inbounds health is false
else for each inbounds we find the container that exposes the pod for the service and if the container state is ready then the inbound is healthy.

> If you specify `httpGet` probe for the Pod, Kuma will generate a special non-MTLs listener and overrides the probe itself in 

Reviewer: Might worth mentioning it works both for liveness and readiness probes.

For example, if we specify the following probe:

```yaml
livenessProbe:
  httpGet:
    path: /metrics
    port: 3001
  initialDelaySeconds: 3
  periodSeconds: 3
```

Kuma will replace it with:

```yaml
livenessProbe:
  httpGet:
    path: /3001/metrics
    port: 9000
  initialDelaySeconds: 3
  periodSeconds: 3
```

Where `9000` is a default virtual probe port, which can be configured in `kuma-cp.config`:

```yaml
runtime:
  kubernetes:
    injector:
      virtualProbesPort: 19001
```
And can also be overwritten in the Pod's annotations:

```yaml
annotations:
  kuma.io/virtual-probes-port: 19001
```

To disable Kuma's probe virtualziation, we can either set it in Kuma's configuration file `kuma-cp.config`:

```yaml
runtime:
  kubernetes:
    injector:
      virtualProbesEnabled: false
```

or in the Pod's annotations:

```yaml
annotations:
  kuma.io/virtual-probes: disabled
```

The same behaviour could be configured using environment variables: 

- `KUMA_RUNTIME_KUBERNETES_VIRTUAL_PROBES_ENABLED=false`
- `KUMA_RUNTIME_KUBERNETES_VIRTUAL_PROBES_ENABLED=19001`

## Universal probes

> On Universal there is no single standard for probing the service. For health checking of the service status on

Reviewer: ```suggestion
On Universal there is no single standard for probing the service. For health checking the inbound health on
```

> Universal Kuma is using Envoy's Health Discovery Service (HDS). Envoy does health checks and reports the status back to Kuma Control Plane.

Reviewer: ```suggestion
Universal Kuma can use Envoy's Health Discovery Service (HDS). Envoy does health checks and reports the status back to Kuma Control Plane.
```

Reviewer: Maybe there should be something like:

In universal you can implement dataplane health in 2 ways:

1. out of bound by having your orchestrator update the dataplane resource using the http api
2. using Envoy HDS

We'll explain how HDS work:

Reviewer: Actually I believe that if you do `1` you might want to describe `Dataplane.Probes` which enables to passthrough mTLS for health checks (that's what k8s does).
